OSDOCS-19890 created zero-trust assembly/modules for multi-cluster (#112274)
openshift/openshift-docs: modules/zero-trust-manager-deploy-east-west.adoc

// Module included in the following assemblies:
//
// * security/zero_trust_workload_identity_manager/zero-trust-manager-mesh-integration-multi-cluster.adoc.

:_mod-docs-content-type: PROCEDURE
[id="zero-trust-manager-deploy-east-west_{context}"]

> = Deploy East-West gateways

Reviewer: Gerund. We haven't moved to imperatives yet.

[role="_abstract"]
Deploy SPIRE-enabled East-West gateways on both clusters using Helm. {SMProductName} uses East-West gateways to connect cluster networks and enable secure cross-cluster communication in a multi-cluster mesh.

.Prerequisites

> * You have completed the "Deploy the Istio custom resource with the federation configuration" steps.

Reviewer: Similar comment for this as previous file. Define the action to be performed and provide the resource in "For more information, see.."

> * The environment variables from the deploy SPIRE procedure are still set in your shell session.

Reviewer: something like this. I am not too sure what it means but please reword it to what's technically correct.
```suggestion
* The environment variables from the SPIRE deployment are set in your shell session.
```

* Federated `ClusterSPIFFEID` resources exist on both clusters.

.Procedure

. Add the Istio Helm repository by running the following commands:
+
[source,terminal]
----
$ helm repo add istio https://istio-release.storage.googleapis.com/charts
----
+
[source,terminal]
----

> $ helm repo update

Reviewer: split steps

----

> . Grant security context constraints (SCC) permissions by running the following commands:

Reviewer: I see only one command
```suggestion
. Grant security context constraints (SCC) permissions by running the following command:
```

done
----

. Install the Istio gateway on Cluster A by running the following command:
+
[source,terminal]
----
$ helm upgrade --install istio-eastwestgateway istio/gateway \
  -n istio-system \
  --set-json 'podAnnotations={"inject.istio.io/templates":"gateway,spireGw"}' \
  --set name=istio-eastwestgateway \
  --set networkGateway="${NETWORK_A}" \
  --kubeconfig="${CLUSTER_A_KUBECONFIG}"
----

. Install the Istio gateway on Cluster B by running the following command:
+
[source,terminal]
----
$ helm upgrade --install istio-eastwestgateway istio/gateway \
  -n istio-system \
  --set-json 'podAnnotations={"inject.istio.io/templates":"gateway,spireGw"}' \
  --set name=istio-eastwestgateway \
  --set networkGateway="${NETWORK_B}" \
  --kubeconfig="${CLUSTER_B_KUBECONFIG}"
----

. Wait for the gateways to complete by running the following command:
+
[source,terminal]
----
$ for KUBECONFIG in "${CLUSTER_A_KUBECONFIG}" "${CLUSTER_B_KUBECONFIG}"; do
  oc wait --for=condition=Available deployment/istio-eastwestgateway \
    --kubeconfig="${KUBECONFIG}" -n istio-system --timeout=300s
done
----

. Create the cross-network Gateway CR on both clusters by running the following command:
+

> The Gateway CR configures the East-West gateway deployment to accept cross-cluster TLS traffic on port 15443 using `AUTO_PASSTHROUGH` mode. This preserves SPIRE-issued certificates for end-to-end mTLS.

Reviewer: This can be written after the code block. This location could be distracting for users.

// Module included in the following assemblies:
//
// * security/zero_trust_workload_identity_manager/zero-trust-manager-mesh-integration-multi-cluster.adoc.

:_mod-docs-content-type: PROCEDURE
[id="zero-trust-manager-deploy-east-west_{context}"]

> = Deploying East-West gateways

Reviewer 2: ```suggestion
= Deploying east-west gateways
```
usually these networking terms are lower case, see the tables here, https://docs.redhat.com/en/documentation/openshift_container_platform/4.22/html/multiple_networks/understanding-multiple-networks#support-matrix-for-udn-nad_understanding-multiple-networks

[role="_abstract"]

> Deploy SPIRE-enabled East-West gateways on both clusters using Helm. {SMProductName} uses East-West gateways to connect cluster networks and enable secure cross-cluster communication in a multi-cluster mesh.

Reviewer 2: ```suggestion
Deploy SPIRE-enabled east-west gateways on both clusters using Helm. {SMProductName} uses east-west gateways to connect cluster networks and enable secure cross-cluster communication in a multi-cluster mesh.
```

+
[source,terminal]
----
$ oc adm policy add-scc-to-user anyuid \
  -z istio-eastwestgateway -n istio-system --kubeconfig="${CLUSTER_A_KUBECONFIG}"
----

. Grant security context constraints (SCC) permissions on Cluster B by running the following command:
+
[source,terminal]
----
$ oc adm policy add-scc-to-user anyuid \
  -z istio-eastwestgateway -n istio-system --kubeconfig="${CLUSTER_B_KUBECONFIG}"
----

. Install the Istio gateway on Cluster A by running the following command:
+
[source,terminal]
----
$ helm upgrade --install istio-eastwestgateway istio/gateway \
  -n istio-system \
  --set-json 'podAnnotations={"inject.istio.io/templates":"gateway,spireGw"}' \
  --set name=istio-eastwestgateway \
  --set networkGateway="${NETWORK_A}" \
  --kubeconfig="${CLUSTER_A_KUBECONFIG}"
----

. Install the Istio gateway on Cluster B by running the following command:
+
[source,terminal]
----
$ helm upgrade --install istio-eastwestgateway istio/gateway \
  -n istio-system \
  --set-json 'podAnnotations={"inject.istio.io/templates":"gateway,spireGw"}' \
  --set name=istio-eastwestgateway \
  --set networkGateway="${NETWORK_B}" \
  --kubeconfig="${CLUSTER_B_KUBECONFIG}"
----

> . Wait for the east-west gateway to become available on Cluster A by running the following command:

Reviewer 2: ```suggestion
. Wait for the East-West gateway to become available on Cluster A by running the following command:
```
use consistent caps throughout; lowercase is likely the preferred style

+
[source,terminal]
----
$ oc wait --for=condition=Available deployment/istio-eastwestgateway \
  --kubeconfig="${CLUSTER_A_KUBECONFIG}" -n istio-system --timeout=300s
----

. Wait for the east-west gateway to become available on Cluster B by running the following command:
+
[source,terminal]
----
$ oc wait --for=condition=Available deployment/istio-eastwestgateway \
  --kubeconfig="${CLUSTER_B_KUBECONFIG}" -n istio-system --timeout=300s
----

> . Create the cross-network Gateway CR on Cluster A by running the following command:

Reviewer 2: ```suggestion
. Create the cross-network `Gateway` custom resource (CR) on Cluster A by running the following command:
```

+
[source,terminal]
----
$ cat <<EOF | oc apply --kubeconfig="${CLUSTER_A_KUBECONFIG}" -f -
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: cross-network-gateway
  namespace: istio-system
spec:
  selector:
    istio: eastwestgateway
  servers:
    - port:
        number: 15443
        name: tls
        protocol: TLS
      tls:
        mode: AUTO_PASSTHROUGH
      hosts:
        - "*.local"
EOF
----

> . Create the cross-network Gateway CR on Cluster B by running the following command:

Reviewer 2: ```suggestion
. Create the cross-network `Gateway CR` on Cluster B by running the following command:
```

+
[source,terminal]
----
$ cat <<EOF | oc apply --kubeconfig="${CLUSTER_B_KUBECONFIG}" -f -
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: cross-network-gateway
  namespace: istio-system
spec:
  selector:
    istio: eastwestgateway
  servers:
    - port:
        number: 15443
        name: tls
        protocol: TLS
      tls:
        mode: AUTO_PASSTHROUGH
      hosts:
        - "*.local"
EOF
----

> The Gateway CR configures the East-West gateway deployment to accept cross-cluster TLS traffic on port 15443 using `AUTO_PASSTHROUGH` mode. This preserves SPIRE-issued certificates for end-to-end mTLS.

Reviewer 2: ```suggestion
+
The `Gateway` CRs configure the East-West gateway deployment to accept cross-cluster TLS traffic on port 15443 using `AUTO_PASSTHROUGH` mode. This preserves SPIRE-issued certificates for end-to-end mTLS.
```
IMO creating the Gateway objects could be a separate module from the other steps

.Verification

> . Verify that the cross-network Gateway exists on both clusters by running the following command:

Reviewer 2: ```suggestion
. Verify that the cross-network `Gateway` CR exists on both clusters by running the following command:
```

+
[source,terminal]
----
$ for KUBECONFIG in "${CLUSTER_A_KUBECONFIG}" "${CLUSTER_B_KUBECONFIG}"; do
  echo "--- $(oc whoami --show-server --kubeconfig="${KUBECONFIG}") ---"
  oc get gateway cross-network-gateway -n istio-system \
    --kubeconfig="${KUBECONFIG}" \
    -o jsonpath='{.spec.servers[0].tls.mode}{"\n"}'
done
----
+
.Example output
+
[source,text]
----
AUTO_PASSTHROUGH
AUTO_PASSTHROUGH
----

> 

Reviewer 2: i would delete all but one line space

----

. Create the cross-network Gateway CR on Cluster B:

.. Create a YAML file that defines the `Gateway` CR on Cluster B:
+
[source,yaml]
----
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: cross-network-gateway
  namespace: istio-system
spec:
  selector:
    istio: eastwestgateway
  servers:
    - port:
        number: 15443
        name: tls
        protocol: TLS
      tls:
        mode: AUTO_PASSTHROUGH
      hosts:
        - "*.local"
----

.. Apply the YAML file on Cluster B by running the following command:
+
[source,terminal]
----
$ oc apply --kubeconfig="${CLUSTER_B_KUBECONFIG}" -f <filename>
----
+
The Gateway CR configures the East-West gateway deployment to accept cross-cluster TLS traffic on port 15443 using `AUTO_PASSTHROUGH` mode. This preserves SPIRE-issued certificates for end-to-end mTLS.

.Verification

. Verify that the cross-network Gateway exists on Cluster A by running the following command:
+
[source,terminal]
----
$ oc get gateway cross-network-gateway -n istio-system \
  --kubeconfig="${CLUSTER_A_KUBECONFIG}" \
  -o jsonpath='{.spec.servers[0].tls.mode}{"\n"}'
----
+

> .Example output

Reviewer 3: @wgabor0427 I think you need to remove the `+` to clear this error.

+
[source,text]
----
AUTO_PASSTHROUGH
----

. Verify that the cross-network Gateway exists on Cluster B by running the following command:
+
[source,terminal]
----
$ oc get gateway cross-network-gateway -n istio-system \
  --kubeconfig="${CLUSTER_B_KUBECONFIG}" \
  -o jsonpath='{.spec.servers[0].tls.mode}{"\n"}'
----
+

> .Example output

Reviewer 3: @wgabor0427 I think you need to remove the + to clear this error.
